juju/juju
# Precheck target controller migrations

Ensure we pass all the facades that are available to the client so that it can verify with the target controller that there is an overlap. This is to ensure that the agent doesn't become wedged if certain API facades aren't available. The source client passes all the facades available to it. It's far better to be conservative with what we send, by sending all of the facades that a subset. The target controller pre-check only checks for a subset. We could add more facades, but we can discuss that in the PR. The facade `CompleteIntersection` code only checks for a partial overlap. If a facade called `Foo`, with versions `1,2,3` on the client is sent to the target controller, and the target controller has version `3,4,5`. We have an intersection of `3`. The migration can continue. This allows us to remove old facades, as long as there is some sort of overlap.

This should still work against 3.x controllers, as we only validate against all facades if the version 3 or greater. Ideally, we'll move all facades up to 3 once that's landed. Allowing us to be a lot more secure.

## Checklist


## QA steps

```sh
$ juju bootstrap lxd src --build-agent
$ juju add-model default
$ juju deploy ubuntu
```

```sh
$ juju bootstrap lxd dst --build-agent
$ juju switch src
```

Watching both controllers in two different terminals:

```sh
$ watch juju status -m src:default
$ watch juju status -m dst:default
```
The latter one won't appear until the model is migrated.

```sh
$ juju migrate src:default dst
```

## Links

**Jira card:** JUJU-4962

**LP Bug:** https://bugs.launchpad.net/juju/+bug/2043986
